### ISSUE TITLE
Extend to diagnosis graph [Bounty 1000€ + Boni]

### ISSUE THREAD

**Issue by 0kSure** [HumanbiOS/telegram_bot_frontend#18]
For example: 
Add Question: "Do you experience difficulties in breathing at the moment" to the medical path. Try to make it as easy as possible for new devs to add new questions. 
Document how new questions can be added.

**Comment by Poolitzer** [comment]
That looks like a fun thing to start with. Do you only want yes/no questions? Do you have a set of questions laying around somewhere already? What about multi language support?

**Comment by 0kSure** [comment]
@Poolitzer We need a Tree (have a look at [networkx](https://networkx.github.io/) for that) as an abstract data structure here for the Q&A bot.
@CypherCrow can help here and wants to make this compatible with Scikit learn/Tensorflow decision trees which we can use to classify/diagnose social/medical. 

As an example, 

![image](https://user-images.githubusercontent.com/61067943/78025008-98bc9200-7359-11ea-8269-95b8a9d36be3.png)


We have a few requirements for the issue to be considered done:
- We need yes/no questions and easy to use numerical user input. 
- Questions must be easy to add on arbitrary vertices [Bonus: adding a question takes under 20 minutes for a beginner dev]. Vertices must be clearly identifiable. 
- Bonus: Graph is plottable 

References:
- NetworkX: https://networkx.github.io/
- Scikit: https://scikit-learn.org/stable/modules/tree.html
- Tensorflow https://www.tensorflow.org/tutorials/estimator/boosted_trees

**Comment by 0kSure** [comment]
Depends on: #25

**Comment by Poolitzer** [comment]
Okay, that sounds great, it would definitely expand my developing horizon. Some questions still remain. 

How do you want to manage multi language support? Each language gets a tree? Or same tree and we translate the strings from the tree?

If we want to translate it, do you want to link it to an online translating platform? Is that something you want to do for bot messages anyway?

You want easy access for beginner devs to add questions to the graph. It might be worth thinking about a "live" question addition in the bot. We would need inline navigation and figure out a nice message represanytation of the graph, maybe even generating images. It would be quite some work to get it implemented, but maybe worth it if "everybody" who wants can add new questions without the need to restart the bot.



Oh and btw, what is supposed to happen with the end result. Forwarded to the medical help group?

**Comment by Poolitzer** [comment]
Okay, that sounds great, it would definitely expand my developing horizon. Some questions still remain. 

How do you want to manage multi language support? Each language gets a tree? Or same tree and we translate the strings from the tree?

If we want to translate it, do you want to link it to an online translating platform? Is that something you want to do for bot messages anyway?

You want easy access for beginner devs to add questions to the graph. It might be worth thinking about a "live" question addition in the bot. We would need inline navigation and figure out a nice message represanytation of the graph, maybe even generating images. It would be quite some work to get it implemented, but maybe worth it if "everybody" who wants can add new questions without the need to restart the bot.



Oh and btw, what is supposed to happen with the end result. Forwarded to the medical help group?

**Comment by 0kSure** [comment]
thx for asking. 

> How do you want to manage multi language support? Each language gets a tree? Or same tree and we translate the strings from the tree? 

I wrote something in #15, I'd make it work for English now and treat the translation as a separate problem. I guess we would have to normalize out the telegram API calls and wrap them with a translator that does another API call for every sent message. 

> You want easy access for beginner devs to add questions to the graph. It might be worth thinking about a "live" question addition in the bot. We would need inline navigation and figure out a nice message representation of the graph, maybe even generating images. It would be quite some work to get it implemented, but maybe worth it if "everybody" who wants can add new questions without the need to restart the bot.

Let me reprioritize here a little, sorry for that. The idea is, that probably more info about cases becomes available, we need a way to add them without refactoring the codebase. So that was my idea here. 
In general, supporting a Graph is the most important thing right now. 

> Oh and btw, what is supposed to happen with the end result. Forwarded to the medical help group?

Yes, that is #20. Would you put those two together?

**Comment by 0kSure** [comment]
I'll update my issue comment to reflect the changes.

**Comment by Poolitzer** [comment]
Okay, I think my questions are answered. For now :evil:


I would suggest implementing a markup translating API call. We didnt do it yet in the other areas, but this part will fundamentally differ on how we "ask questions" from the other areas anyway, so I would say it feels appropriate. That way, it will be easier to implement multi language support later, at least for the graph.

**Comment by 0kSure** [comment]
One Json file per language, where a node is defined as:
`(node_id= 1, Label ="How old are you?"), edges = [(1,2), id = "OptionA", (1,3), id = "OptionB"]).` 

From the JSON, generate StateObjects, they can be really generic and we right now only need one multiple-choice Object, with yes/no as a special case. 

The StateObjects are initialized as a NetworkX graph with a Node being (node_id, StateObject). 
The Bot wraps the Graph and does path traversal, e.g. starts with Node0, State0. User input determines the edge that is traversed. An edge represents a state transition @d-Rickyy-b this is probably where the batch of the logic you talked about will be in. Node_2 is than the next State Object.
